File: haskell-server/src/BoardGame/Common/Domain/PiecePoint.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE StandaloneDeriving #-}

module BoardGame.Common.Domain.PiecePoint (
    PiecePoint(..)
  , pointLetter
)
where

import GHC.Generics (Generic)
import Control.DeepSeq (NFData)
import Data.Aeson (FromJSON, ToJSON)
import BoardGame.Common.Domain.Piece
import qualified BoardGame.Common.Domain.Piece as Piece
import Bolour.Plane.Domain.Point (Point)

-- | A piece located at a point on the plane.
data PiecePoint = PiecePoint {
    piece :: Piece    -- ^ The piece.
  , point :: Point    -- ^ The position of the piece on the plane.
} deriving (Eq, Show, Generic, NFData)

instance FromJSON PiecePoint
instance ToJSON PiecePoint

pointLetter :: PiecePoint -> Char
pointLetter PiecePoint {piece} = Piece.value piece

File: scala-server/boardgamecommon/src/main/scala/com/bolour/boardgame/scala/common/domain/PiecePoint.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */
package com.bolour.boardgame.scala.common.domain

import com.bolour.plane.scala.domain.Point

/**
  * Combination of a piece and its location on the board.
  */
case class PiecePoint(piece: Piece, point: Point)
